File: problems_1000-2000/1365-HowManyNumbersAreSmallerThantheCurrentNumber/python/solution.py
class Solution:
    def smallerNumbersThanCurrent(self, nums: list[int]) -> list[int]:
        max_val = max(nums)
        count = [0] * (max_val + 1)
        for num in nums:
            count[num] += 1
        for i in range(1, max_val + 1):
            count[i] += count[i - 1]
        result = []
        for num in nums:
            if num == 0:
                result.append(0)
            else:
                result.append(count[num - 1])
        return result
        
if __name__ == '__main__':
    nums = [8,1,2,2,3]
    s = Solution()
    print(s.smallerNumbersThanCurrent(nums))

File: problems_1000-2000/1365-HowManyNumbersAreSmallerThantheCurrentNumber/go/solution.go
package main

import "fmt"

func smallerNumbersThanCurrent(nums []int) []int {

	maxVal := 0
	for _, num := range nums {
		if num > maxVal {
			maxVal = num
		}
	}

	count := make([]int, maxVal+1)

	for _, num := range nums {
		count[num]++
	}

	for i := 1; i <= maxVal; i++ {
		count[i] += count[i-1]
	}

	result := make([]int, len(nums))
	for i, num := range nums {
		if num == 0 {
			result[i] = 0
		} else {
			result[i] = count[num-1]
		}
	}

	return result
}

func main() {
	nums := []int{8, 1, 2, 2, 3}
	result := smallerNumbersThanCurrent(nums)
	fmt.Println(result)
}
